src/net/sourceforge/kolmafia/KoLAdventure.java

```diff
@@ -1816,6 +1855,26 @@ public boolean prepareForAdventure() {
       return true;
     }
 
+    if (this.zone.equals("FantasyRealm")) {
+      if (!InventoryManager.hasItem(FANTASY_REALM_GEM)) {
```

Comment:
This is reasonable, but a counterargument would be how the Astral adventures are intended to work:

If you ask to take such-and-such a trip, if you have not already consumed the mushroom and chosen, we will use the mushroom and trigger the initial choice adventure to choose the trip. At that point, you are all set up and can go right into the chosen zone.

Here, the "get a GEM (while choosing a hat) and equip it" is, sort of, the equivalent of "use a mushroom and make a request to choose the trip". The difference is, the "type of hat" is not implied by the adventure ID.

In PrepareForAdventure, the user is saying "suit me up! I am ready to go in!"

Now, perhaps if they wanted a particular hat to actually wear in this zone, presumably they would have picked it first - and got the GEM. If they didn't do that, if they have set a "default" for the hat picking choice, perhaps we should automate it for them and get the GEM as well. Is the hat choice GUI-configurable?

I'm leaning towards that, but don't feel strongly about it.

Reply:
I started doing this but it would actually make it quite complicated - now our canAdventure function needs to know what locations *would* become available if you were to complete the Welcome choice and return true for them, but that pref is only set after visiting. It's possible, I'd split out the function that currently works that out and put it in FantasyRealmRequest or something. But I think it's probably beyond the scope of this change, and I feel like doing a bit more breadth before this depth. What do you think?

Reply:
I didn't know that. I had the impression that the unlocked zones persisted, not that we learned them by looking at the map.

Yeah, this is fine. Once we have canAdventure and prepareForAdventure  in-synch - the former checks for everything the latter will provide - then we can consider going deeper into individual zones.